### DISCUSSION TITLE
fetch_or instead of cas-loop in mutex code

### DISCUSSION THREAD

**Discussion by MBkkt** [abseil/abseil-cpp#1577] [Ideas]
I noticed in [that commit](https://github.com/abseil/abseil-cpp/commit/f3760b4d3b2773d1cb8e9ddda29dc9bce386d540) function `AtomicSetBits` and thought it can be better written like this, because on armv8 fetch_or is "single" instruction

```cpp
    intptr_t v;
    do {
        v = pv->load(std::memory_order_relaxed);
        if ((v & bits) != 0) {
            return false;
        }
    } while ((v & wait_until_clear) != 0);
    return (pv->fetch_or(bits, std::memory_order_release) & bits) == 0;
```

It's not the same logic (wait_until_clear not checked after achieved), but it seems like it doesn't matter and also can be used.

From other side clang on x86 cannot put condition with early-exit in the "fetch_or" cas-loop

**Comment by derekmauro** [comment]
As you said, this isn't quite the same thing. I'm not sure that this is worth optimizing since it is only used in a debug path which is already going to be slow.

/cc @dvyukov

**Reply by MBkkt** [reply]
Is it debug only? Ok, I didn't notice

**Comment by dvyukov** [comment]
>     } while ((v & wait_until_clear) != 0);
>    return (pv->fetch_or(bits, std::memory_order_release) & bits) == 0;

This looks broken. The flag may be set in between the check and the fetch_or.

**Reply by MBkkt** [reply]
I thought wait_until_clear cannot be set after it was cleared.
If it's not true, ofc, thanks
